src/iarc7_motion/state_monitor.py

```diff
@@ -0,0 +1,164 @@
+
+        self._current_velocity_sub = rospy.Subscriber(
+            '/odometry/filtered', Odometry,
+            self._recieve_drone_odometry)
+
+        self._drone_arm_status = rospy.Subscriber(
+            '/fc_status', FlightControllerStatus, 
+            self._receive_arm_status)
+
+        try:
+            # task timeout time to decelerate
+            self._task_timeout_deceleration_time = rospy.Duration(
+                rospy.get_param('~task_timeout_deceleration_time'))
+            # minimum safe height to manuever at
+            self._MIN_MANEUVER_HEIGHT = rospy.get_param('~min_maneuver_height')
+
+        except KeyError as e:
+            rospy.logerr('Could not lookup a parameter for State Monitor')
+            raise
+
+    # checks task transitions before executing it
+    def check_transition(self, task):
+        with self._lock: 
+            passed = False
+
+            if self._waiting_on_takeoff: 
+                passed = isinstance(task, TakeoffTask)
+            elif isinstance(task, LandTask):
+                self._waiting_on_takeoff = True
+                passed = True
+            elif self._waiting_on_recovery:
+                passed = isinstance(task, HeightRecoveryTask)
+            elif self._below_min_manuever_height():
+                passed = isinstance(task, TakeoffTask) or isinstance(task, HeightRecoveryTask)
+            else: 
+                passed = True
+
+            if passed:
+                if isinstance(task, TakeoffTask):
+                    self._waiting_on_takeoff = False
+                elif isinstance(task, BlockRoombaTask) or isinstance(task, HitRoombaTask):
+                    self._waiting_on_recovery = True
+                elif isinstance(task, HeightRecoveryTask):
+                    self._waiting_on_recovery = False
+
+            rospy.loginfo("Waiting on Takeoff: " + str(self._waiting_on_takeoff))
+            rospy.loginfo("Waiting on Recovery: " + str(self._waiting_on_recovery))
+
+            return passed
+
+    # checks to see if drone above min manuever height
+    def _below_min_manuever_height(self):
+        with self._lock:
+            if self._drone_odometry is None:
+                return True
+            else:
+                return (self._drone_odometry.pose.pose.position.z 
+                            < self._MIN_MANEUVER_HEIGHT)
+
+    # fills out the rest of Intermediary State for the task
+    def fill_out_transition(self, state):
+        while ((self._drone_odometry is None or self._roombas is None 
+                or self._arm_status is None) and not rospy.is_shutdown()):
+            pass
+        with self._lock:
+            state.drone_odometry = self._drone_odometry
+            state.roombas = self._roombas
+            state.obstacles = self._obstacles
+            state.arm_status = self._arm_status
+            return state
+    
+    # Handles no task running timeouts
+    def get_timeout(self, last_twist):
```

Comment:
I don't think this belongs here. Its a safety feature of high_level_motion_controller.py in my opinion. This class should handle the transition just like any other task.

Reply:
The thought was that if we wanted to avoid obstacles when timing out or have a more robust timeout system, then we would need to be plugged into current state information, which the state monitor is already.